### ISSUE TITLE
Data corruption depending on input file name

### ISSUE THREAD

**Issue by KirillKryukov** [cobilab/geco3#1]
I noticed that GeCo3 sometimes corrupts data. I.e., after compression + decompression, the decompressed sequence is different from the original one.

I uploaded a small test dataset: http://kirill.med.u-tokai.ac.jp/data/temp/GeCo3-repro-1-test-data.seq.gz (1.4 MB, gzipped to 0.4 MB).

Steps to reproduce (on Linux):

```
cd /tmp
mkdir GeCo3-repro-1
cd GeCo3-repro-1
git clone https://github.com/cobilab/geco3.git
cd geco3/src
make
cd ../..
wget http://kirill.med.u-tokai.ac.jp/data/temp/GeCo3-repro-1-test-data.seq.gz
gzip -dc GeCo3-repro-1-test-data.seq.gz >GeCo3-repro-1-test-data.seq
geco3/src/GeCo3 -l 1 GeCo3-repro-1-test-data.seq
geco3/src/GeDe3 GeCo3-repro-1-test-data.seq.co
cmp GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de
```

Console output of the last few steps:

```
lr: 0.03, hs: 40
xs: 35
Total bytes: 347972 (339.8 KB), 1.924 bpb, 1.924 bps w/ no header, Normalized Dissimilarity Rate: 0.962092
Spent 2.37155 sec.
xs: 35
Spent 2.05411 sec.
GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de differ: byte 31, line 1
```

The decompressed file "GeCo3-repro-1-test-data.seq.de" has correct size, but completely different sequence. This problem occurs only with some data, and only at some compression levels.

A little experimentation revealed that curiously this issue seems to be sensitive to the input file name.

The following Perl script tries GeCo3 with levels from 1 to 7, verifying the output of each run. Then it renames the file to "1.seq" and again tests GeCo3 levels 1 to 7:

```
print "\nInput is named \"GeCo3-repro-1-test-data.seq\"\n";
for (my $l = 1; $l <= 7; $l++)
{
    system("geco3/src/GeCo3 -l $l GeCo3-repro-1-test-data.seq >/dev/null 2>&1");
    system('geco3/src/GeDe3 GeCo3-repro-1-test-data.seq.co >/dev/null 2>&1');
    my $r = `cmp GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de`;
    print "$l: ", (($r eq '') ? "OK\n" : $r);
    unlink 'GeCo3-repro-1-test-data.seq.co';
    unlink 'GeCo3-repro-1-test-data.seq.de';
}
rename 'GeCo3-repro-1-test-data.seq', '1.seq';
print "\nNow input is renamed to \"1.seq\"\n";
for (my $l = 1; $l <= 7; $l++)
{
    system("geco3/src/GeCo3 -l $l 1.seq >/dev/null 2>&1");
    system('geco3/src/GeDe3 1.seq.co >/dev/null 2>&1');
    my $r = `cmp 1.seq 1.seq.de`;
    print "$l: ", (($r eq '') ? "OK\n" : $r);
    unlink '1.seq.co';
    unlink '1.seq.de';
}
```

Steps to download and run the above script:
```
cd /tmp
mkdir GeCo3-repro-1-part-2
cd GeCo3-repro-1-part-2
git clone https://github.com/cobilab/geco3.git
cd geco3/src
make
cd ../..
wget http://kirill.med.u-tokai.ac.jp/data/temp/GeCo3-repro-1-test-data.seq.gz
gzip -dc GeCo3-repro-1-test-data.seq.gz >GeCo3-repro-1-test-data.seq
wget http://kirill.med.u-tokai.ac.jp/data/temp/GeCo3-repro-1-test.pl
/usr/bin/env perl GeCo3-repro-1-test.pl
```

The output of this step:

```
Input is named "GeCo3-repro-1-test-data.seq"
1: GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de differ: byte 31, line 1
2: OK
3: OK
4: GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de differ: byte 40, line 1
5: GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de differ: byte 40, line 1
6: GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de differ: byte 40, line 1
7: OK

Now input is renamed to "1.seq"
1: OK
2: 1.seq 1.seq.de differ: byte 29, line 1
3: 1.seq 1.seq.de differ: byte 17, line 1
4: OK
5: OK
6: OK
7: 1.seq 1.seq.de differ: byte 59, line 1
```

Leaking input file name into other data structures may be indication of larger memory safety problems. It may be good idea to review the code for memory safety issues. May be you can consider using static analysis tools such as CoverityScan.

Test machine: Ubuntu 18.04.1 LTS, dual Xeon E5-2643v3.

Let me know if you need any additional information or testing.

**Comment by miltondts** [comment]
Following the exact steps provided I was not able to reproduce the problem (even with the perl script).
Our test environment is:
Linux kernel used: 5.6.15-1-MANJARO
gcc (GCC) 10.1.0

Can you provide us the gcc version you are using? Also, can you verify if modifying line 8 of Makefile to be `CFLAGS = -Wall -O2 $(CPLP)` changes the behavior?

Regarding memory corruption, valgrind only shows some missing frees. Which we will fix, but should have no impact on this issue. Once we can reproduce this issue, will check with CoverityScan.

EDIT:
We also tested with docker:

```
docker run -it ubuntu:18.04
apt update
apt install wget git build-essential

# steps to reproduce
cd /tmp
mkdir GeCo3-repro-1
cd GeCo3-repro-1
git clone https://github.com/cobilab/geco3.git
cd geco3/src
make
cd ../..
wget http://kirill.med.u-tokai.ac.jp/data/temp/GeCo3-repro-1-test-data.seq.gz
gzip -dc GeCo3-repro-1-test-data.seq.gz >GeCo3-repro-1-test-data.seq
geco3/src/GeCo3 -l 1 GeCo3-repro-1-test-data.seq
geco3/src/GeDe3 GeCo3-repro-1-test-data.seq.co
cmp GeCo3-repro-1-test-data.seq GeCo3-repro-1-test-data.seq.de
```

This does indeed show an error. Changing the compilation to O2 solves the problem. 

We will investigate if this is a problem with gcc or if it is a problem with the program.

**Comment by KirillKryukov** [comment]
GCC 7.5.0. Thanks for quick response, and I'm looking forward to your fix.

**Comment by miltondts** [comment]
The issue doesn't occur at least with the following compilers:
clang: 6.0.0 and 10.0.0
gcc: 10.1.0

GCC 7.5.0 has this issue but only with the flag -march=native.

Valgrind and gcc sanitizers don't detect any undefined behavior, stack smashing, alignment issues or other memory corruption.
CoverityScan doesn't work on the kernels we have access to. It is mentioned on the website that:
"Linux users on kernel version 4.8.x and newer will need to apply a sysctl parameter to support our binaries." However even with the suggested workaround no analysis files are produced.


Until we can pinpoint the source of the problem, we recommend as a temporary workaround (when compiling with GCC 7.5) to edit [this line](https://github.com/cobilab/geco3/blob/0716e19718a3c376071522a0047d01bcec390cd9/src/Makefile#L8) to be the following `CFLAGS = -Wall -Ofast  $(CPLP)`

**Comment by miltondts** [comment]
We have push a code refactor to prevent the issue on GCC 7.5.
The commit is [here](https://github.com/cobilab/geco3/commit/8ba29984a4ae3f847ca6204d4a053eb6f0c22b11).
The commit message has more details on our findings.
Once we have more time we will do a more complete investigation.

**Comment by KirillKryukov** [comment]
The refactored geco3 seems working so far, so I can continue my testing. Thanks for quick work. And I agree that it's a good idea to do a more complete investigation. There may be some uninitialized memory, or other problem. The connection of this floating point math to input file name is still a mystery.

**Comment by miltondts** [comment]
@KirillKryukov just to let you know that the fix is now in master. I've validated it with the version of code that produced the problem, but I kept the refactor as it doesn't change the results and because there is one less loop it might be slightly faster.

The problem is that -Ofast leads to undefined behavior and this causes slightly different results in the compressor and decompressor that where then compounded because the network outputs are used to generate network inputs. Originally, -Ofast was used because we where having difficulties with gcc's automatic vectorization. We have now changed to -O3 without much difference in performance and without the problems that -Ofast brings.
